projects/plugins/boost/app/lib/minify/class-cleanup-stored-paths.php

```diff
@@ -0,0 +1,149 @@
+	 *
+	 * @var int
+	 */
+	private $max_options_to_process = 50;
+
+	/**
+	 * The key of the option that stores the ID of the last processed option.
+	 *
+	 * @var string
+	 */
+	private $last_processed_option_key = 'jetpack_boost_cleanup_concat_paths_last_processed_option_id';
+
+	/**
+	 * Whether to schedule a followup cleanup.
+	 *
+	 * @var bool
+	 */
+	private $should_schedule_followup = false;
+
+	/**
+	 * Schedules the start of the cleanup.
+	 */
+	public static function setup_schedule() {
+		if ( false === wp_next_scheduled( 'jetpack_boost_minify_cron_cleanup_concat_paths' ) ) {
+			wp_schedule_event( time(), 'daily', 'jetpack_boost_minify_cron_cleanup_concat_paths' );
+		}
+	}
+
+	/**
+	 * Hooks the callbacks for the cleanup.
+	 */
+	public static function add_cleanup_actions() {
+		add_action( 'jetpack_boost_minify_cron_cleanup_concat_paths', array( __CLASS__, 'run_cleanup' ) );
+		add_action( 'jetpack_boost_minify_cron_cleanup_concat_paths_followup', array( __CLASS__, 'run_cleanup' ) );
+	}
+
+	/**
+	 * Clears the cleanup schedules.
+	 */
+	public static function clear_schedules() {
+		wp_unschedule_hook( 'jetpack_boost_minify_cron_cleanup_concat_paths' );
+		wp_unschedule_hook( 'jetpack_boost_minify_cron_cleanup_concat_paths_followup' );
+	}
+
+	/**
+	 * Runs the cleanup for the first batch,
+	 * and if there are more entries to process,
+	 * schedules the cleanup for the next batch.
+	 */
+	public static function run_cleanup() {
+		$cleanup      = new Cleanup_Stored_Paths();
+		$can_continue = $cleanup->cleanup_stored_paths_batch();
+		if ( ! $can_continue ) {
+			return;
+		}
+
+		if ( ! $cleanup->should_schedule_followup ) {
+			return;
+		}
+
+		// 'batch' arg is necessary, to tell WP that this is a unique event
+		// and allow it to be run within 10 minutes of the last.
+		// See https://developer.wordpress.org/reference/functions/wp_schedule_single_event/#description
+		if ( ! wp_next_scheduled( 'jetpack_boost_minify_cron_cleanup_concat_paths_followup', array( 'batch' ) ) ) {
+			wp_schedule_single_event( time(), 'jetpack_boost_minify_cron_cleanup_concat_paths_followup', array( 'batch' ) );
+		}
+	}
+
+	/**
+	 * Cleans up expired stored paths.
+	 *
+	 * @return bool True if there are more entries to process, false if not.
+	 */
+	public function cleanup_stored_paths_batch() {
+		$stored_paths = $this->get_stored_paths();
+		if ( ! $stored_paths ) {
+			// Cleanup after the cleanup.
+			delete_option( $this->last_processed_option_key );
+			return false;
+		}
+
+		// Used to tell the cleanup to skip the entries that were checked in the previous run.
+		// Avoids processing the same entries over and over again.
+		$last_processed_option_id = false;
+
+		foreach ( $stored_paths as $option ) {
+			$value = maybe_unserialize( $option['option_value'] );
+			if ( ! is_array( $value ) ) {
+				continue;
+			}
+
+			if ( $value['expire'] <= time() ) {
+				delete_option( $option['option_name'] );
```

Comment:
should we do some type of clean up for the existing year-long transients? If the value is either less than the current time (truly expired) or if the value is more than... 45 days in the future?

With sufficient inline documentation to explain why we care about that...

Reply:
We considered cleaning up transients that are year-long. Ultimately, we decided against cleaning them up, as removing them might do more harm than good.

The harm _might_ come from cached pages. If we were to delete such transients, cached pages referencing the deleted hashes will break (as we'll also be cleaning up files belonging to those hashes), because there's no data to reconstruct the file from.

Since such options aren't auto-loaded, they would get cleaned eventually. We'll also be updating the Boost uninstall to use a direct DB query, which should be way faster than the regular delete_option.